src/models/mSeriesChallenge.ts

```diff
@@ -0,0 +1,287 @@
+  ID = 'id',
+  StartAt = 'startAt',
+  EndAt = 'endAt',
+  CreatedAt = 'createdAt'
+}
+
+export interface MSeriesChallengeData {
+  id: number
+  joinCode: string
+  name: string
+  userId: number
+  isPublic: boolean
+  userLimit: number | null
+  challengeType: MSeriesChallengeType
+  startAt: string
+  endAt: string | null
+  focus: MSeriesChallengeFocus
+  goal: number
+  isCompleted: boolean
+  isJoined: boolean
+}
+
+export interface MSeriesChallengeResponse extends AuthenticatedResponse {
+  mSeriesChallenge: MSeriesChallengeData
+}
+
+export interface MSeriesChallengeListResponse extends AuthenticatedResponse {
+  mSeriesChallenges: MSeriesChallengeData[]
+  mSeriesChallengesMeta: MSeriesChallengeListResponseMeta
+}
+
+export interface MSeriesChallengeListResponseMeta extends ListMeta {
+  from: string | undefined
+  to: string | undefined
+  isCompleted: boolean
+  relationship: MSeriesChallengeRelationship
+  sort: MSeriesChallengeSorting
+}
+
+export class BaseMSeriesChallenge extends Model {
+  protected _mSeriesChallengeData: MSeriesChallengeData
+
+  constructor (mSeriesChallengeData: MSeriesChallengeData, sessionHandler: SessionHandler) {
+    super(sessionHandler)
+    this._mSeriesChallengeData = mSeriesChallengeData
+  }
+
+  protected setMSeriesChallenge (mSeriesChallengeData: MSeriesChallengeData) {
+    this._mSeriesChallengeData = mSeriesChallengeData
+  }
+
+  async reload () {
+    const { mSeriesChallenge } = await this.action('mSeriesChallenge:show', { id: this._mSeriesChallengeData.id, userId: this.sessionHandler.userId }) as MSeriesChallengeResponse
+    this.setMSeriesChallenge(mSeriesChallenge)
+    return this
+  }
+
+  get id () {
+    return this._mSeriesChallengeData.id
+  }
+
+  get joinCode () {
+    return this._mSeriesChallengeData.joinCode
+  }
+
+  get userId () {
+    return this._mSeriesChallengeData.userId
+  }
+
+  get name () {
+    return this._mSeriesChallengeData.name
+  }
+
+  get isPublic () {
+    return this._mSeriesChallengeData.isPublic
+  }
+
+  get userLimit () {
+    return this._mSeriesChallengeData.userLimit
+  }
+
+  get challengeType () {
+    return this._mSeriesChallengeData.challengeType
+  }
+
+  get startAt () {
+    return new Date(this._mSeriesChallengeData.startAt)
+  }
+
+  get endAt () {
+    return this._mSeriesChallengeData.endAt !== null ? new Date(this._mSeriesChallengeData.endAt) : null
+  }
+
+  get focus () {
+    return this._mSeriesChallengeData.focus
+  }
+
+  get goal () {
+    return this._mSeriesChallengeData.goal
+  }
+
+  get isCompleted () {
+    return this._mSeriesChallengeData.isCompleted
+  }
+
+  get isJoined () {
+    return this._mSeriesChallengeData.isJoined
+  }
+
+  get mSeriesChallengeData (): MSeriesChallengeData {
+    return { ...this._mSeriesChallengeData }
+  }
+}
+
+// non-exportable class that contains methods shared between JoinedMSeriesChallenge and PrivilegedMSeriesChallenge
+class AccessibleMSeriesChallenge extends BaseMSeriesChallenge {
```

Comment:
This needs to exported for type checking.

This can also be an `abstract` class.